src/unix/syscall.c

```diff
@@ -748,36 +646,54 @@ static int file_type_from_tuple(tuple n)
 {
     if (is_dir(n))
         return FDESC_TYPE_DIRECTORY;
+    else if (is_symlink(n))
+        return FDESC_TYPE_SYMLINK;
     else if (is_special(n))
         return FDESC_TYPE_SPECIAL;
     else
         return FDESC_TYPE_REGULAR;
 }
 
+static int dt_from_tuple(tuple n)
+{
+    if (is_dir(n))
+        return DT_DIR;
+    else if (is_symlink(n))
+        return DT_LNK;
+    else
+        return DT_REG;
+}
+
 sysreturn open_internal(tuple cwd, const char *name, int flags, int mode)
 {
     heap h = heap_general(get_kernel_heaps());
     unix_heaps uh = get_unix_heaps();
-    tuple n = resolve_cstring(cwd, name);
+    tuple n;
+    tuple parent;
+    int ret;
 
+    if (flags & O_NOFOLLOW) {
+        ret = resolve_cstring(cwd, name, &n, &parent);
+        if (!ret && is_symlink(n) && !(flags & O_PATH)) {
+            ret = -ELOOP;
+        }
+    } else {
+        ret = resolve_cstring_follow(cwd, name, &n, &parent);
+    }
     if ((flags & O_CREAT)) {
-        if (n && (flags & O_EXCL)) {
+        if (!ret && (flags & O_EXCL)) {
             thread_log(current, "\"%s\" opened with O_EXCL but already exists", name);
             return set_syscall_error(current, EEXIST);
-        } else if (!n) {
-            fs_status fs = filesystem_creat(current->p->fs, cwd, name, mode);
-            if (fs != FS_STATUS_OK)
-                return sysreturn_from_fs_status(fs);
-
-            /* XXX We could rearrange calls to return tuple instead of
-               status; though this serves as a sanity check. */
-            n = resolve_cstring(cwd, name);
+        } else if ((ret == -ENOENT) && parent) {
+            n = filesystem_creat(current->p->fs, parent,
```

Comment:
Error check here?

Reply:
I will add a check for n != 0; checking for I/O errors during file creation (which wasn't being done even before this commit) would require more work, since it's an async operation.